stock_batch_picking/models/stock_batch_picking.py

```diff
@@ -0,0 +1,213 @@
+        'stock.pack.operation',
+        string='Related pack operations',
+        compute='_compute_pack_operation_pack_ids',
+        inverse=lambda *args, **kwargs: None,
+        states={'done': [('readonly', True)], 'cancel': [('readonly', True)]}
+    )
+
+    @api.depends('picking_ids')
+    def _compute_move_ids(self):
+        for batch in self:
+            batch.move_ids = batch.picking_ids.mapped("move_lines")
+
+    @api.depends('picking_ids')
+    def _compute_pack_operation_ids(self):
+        for batch in self:
+            batch.pack_operation_ids = batch.picking_ids.mapped(
+                'pack_operation_ids'
+            )
+
+    @api.depends('picking_ids')
+    def _compute_pack_operation_product_ids(self):
+        for batch in self:
+            batch.pack_operation_product_ids = batch.picking_ids.mapped(
+                'pack_operation_product_ids'
+            )
+
+    @api.depends('picking_ids')
+    def _compute_pack_operation_pack_ids(self):
+        for batch in self:
+            batch.pack_operation_pack_ids = batch.picking_ids.mapped(
+                'pack_operation_pack_ids'
+            )
+
+    def get_not_empties(self):
+        """ Retourne all batches in this recordset
+        for which picking_ids is not empty.
+
+        :raise UserError: If all batches are empty.
+        """
+        if not self.mapped('picking_ids'):
+            if len(self) == 1:
+                message = _('This Batch has no picking')
+            else:
+                message = _('These Batches have no picking')
+
+            raise UserError(message)
+
+        return self.filtered(lambda b: len(b.picking_ids) != 0)
+
+    def verify_state(self, expected_state=None):
+        """ Check if batches states must be changed based on pickings states.
+
+        If all pickings are canceled, batch must be canceled.
+        If all pickings are canceled or done, batch must be done.
+        If all pickings are canceled or done or *expected_state*,
+            batch must be *expected_state*.
+
+        :return: True if batches states has been changed.
+        """
+        expected_states = {'done', 'cancel'}
+        if expected_state is not None:
+            expected_states.add(expected_state)
+
+        all_good = True
+        for batch in self.filtered(lambda b: b.state not in expected_states):
+            states = set(batch.mapped('picking_ids.state'))
+            if not states or states == {'cancel'}:
+                batch.state = 'cancel'
+            elif states == {'done'} or states == {'done', 'cancel'}:
+                batch.state = 'done'
+
+            elif states.issubset(expected_states):
+                batch.state = expected_state
+
+            else:
+                all_good = False
+
+        return all_good
+
+    @api.multi
+    def action_cancel(self):
+        """ Call action_cancel for all batches pickings
+        and set batches states to cancel too.
+        """
+        if not self.picking_ids:
```

Comment:
It is possible for `self` to be multiple records here, so `self.picking_ids` is not always valid. I recommend either looping the recordset then operating on that, or a `self.ensure_one()`